[FLINK-24098] Document FLIP-147 capabiliites and limitations (#17135)
apache/flink: docs/content/docs/dev/datastream/fault-tolerance/checkpointing.md

>   - *unaligned checkpoints*: You can enable [unaligned checkpoints]({{< ref "docs/ops/state/unaligned_checkpoints" >}}) to greatly reduce checkpointing times under backpressure. Only works for exactly-once checkpoints and with number of concurrent checkpoints of 1.
> 
>   - *checkpoints with finished tasks*: You can enable an experimental feature to continue performing checkpoints even if parts of the DAG have finished processing all of their records. Before doing so please read through the [important considerations](#checkpointing-with-parts-of-the-graph-finished).
> 

Reviewer: ```suggestion
  - *unaligned checkpoints*: You can enable [unaligned checkpoints]({{< ref "docs/ops/state/unaligned_checkpoints" >}}) to greatly reduce checkpointing times under backpressure. This only works for exactly-once checkpoints and with a number of concurrent checkpoints of 1.

  - *checkpoints with finished tasks*: You can enable an experimental feature to continue performing checkpoints even if parts of the DAG have finished processing all of their records. Before doing so, please read through some [important considerations](#checkpointing-with-parts-of-the-graph-finished).

```

// sets the checkpoint storage where checkpoint snapshots will be written
env.getCheckpointConfig.setCheckpointStorage("hdfs:///my/checkpoint/dir")

// enable checkpointing with finished tasks
val config = new Configuration()
config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true)

> env.configure(config)

Reviewer 2: What is the python way to enable this feature?

Author: https://issues.apache.org/jira/browse/FLINK-24199

Reviewer 2: I encourage you to try add it yourself, sometimes it's not that difficult:
https://github.com/apache/flink/pull/17100/commits/e83081e3d734448f232e6f676d71c46a44c4081e

Here it might be a bit more tricky with using {{Configuration}}, but maybe there is already an example how to do it that you can just copy/paste/rename.

Author: I did add methods to the Python API in the past on multiple occasions. Here, I'd really prefer not to deal with it, as it deals with `Configuration`. 

1. First of all I am not able to run the python tests. Configuring the IDE to make them run, would take me quite some time.
2. In Java/Scala we use a type safe version of configuration `ReadableConfig`. I don't know how we should solve that in python.

Reviewer 2: Ok, I get it.

Starting from Flink 1.14 it is possible to continue performing checkpoints even if parts of the job graph have finished processing all data, because it was a bounded source. The feature must be enabled
via a feature flag:

```java
Configuration config = new Configuration();
config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true);
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
```

Once tasks/subtasks are finished they don't contribute to the checkpoints any longer. It is an
important observation that puts certain requirements on the implementation of any custom operators
or UDFs. In order to support checkpointing with tasks that finish we adjusted the [task lifecycle]({{< ref "docs/internals/task_lifecycle" >}})
and introduced the {{< javadoc file="org/apache/flink/streaming/api/operators/StreamOperator.html#finish--" name="StreamOperator#finish" >}}
method. The method is expected to be a clear cutoff point for flushing any remaining buffered state.
All checkpoints taken after the `finish` method has been called should not contain any significant state that is required

> after a restore. What does it mean in details?

Reviewer 2: This and later `discarded` is a bit confusing, as the finished subtasks still can/should have a state pointing to side effects (2pc). It should be phrased that after calling `finish()` state should be empty except of pointers to commit side effects, like external transactions.

Author: I tried rephrasing it.

Please note that records in flight in the loop edges (and the state changes associated with them) will be lost during failure.

## Checkpointing with parts of the graph finished *(BETA)*

> Starting from Flink 1.14 it is possible to continue performing checkpoints even if parts of the job graph have finished processing all data, because it was a bounded source. The feature must be enabled

Author: First of all, I agree, this sentence is not very clear. I would not use `all` / `only`, because this way the reader might be confused that it works iff the job graph contains exclusively bounded sources.

I will rephrase it to: 
> Starting from Flink 1.14 it is possible to continue performing checkpoints even if parts of the job
graph have finished processing all data, which might happen if it contains bounded sources. The
WDYT?

```

Once tasks/subtasks are finished they don't contribute to the checkpoints any longer. It is an
important observation that puts certain requirements on the implementation of any custom operators
or UDFs. In order to support checkpointing with tasks that finish we adjusted the [task lifecycle]({{< ref "docs/internals/task_lifecycle" >}})
and introduced the {{< javadoc file="org/apache/flink/streaming/api/operators/StreamOperator.html#finish--" name="StreamOperator#finish" >}}
method. The method is expected to be a clear cutoff point for flushing any remaining buffered state.
All checkpoints taken after the `finish` method has been called should be in most cases empty and
shouldn't contain any buffered data, as there will be no way to emit this data. One notable
exception is if your operator has some pointers to transactions in external systems, for example in
order to implement the exactly-once semantic. In such a case, checkpoints taken after invoking `finish()`
method should keep a pointer to the last transaction(s) that will be committed in the final checkpoint
before the operator is closed. A good built-in example of this are exactly-once sinks and the 

> `TwoPhaseCommitSinkFunction`. What does it mean in more details?

Author: My idea is to have it there to make it less dull and catch the readers attention again for another important section. I don't have strong attachment to the sentence though ;)

Please note that records in flight in the loop edges (and the state changes associated with them) will be lost during failure.

> ## Checkpointing with parts of the graph finished *(BETA)*
> 
> Starting from Flink 1.14 it is possible to continue performing checkpoints even if parts of the job
> graph have finished processing all data, which might happen if it contains bounded sources. The
> feature must be enabled via a feature flag:
> 
> ```java
> Configuration config = new Configuration();
> config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true);
> StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
> ```
> 
> Once tasks/subtasks are finished they don't contribute to the checkpoints any longer. It is an
> important observation that puts certain requirements on the implementation of any custom operators
> or UDFs. In order to support checkpointing with tasks that finish we adjusted the [task lifecycle]({{< ref "docs/internals/task_lifecycle" >}})
> and introduced the {{< javadoc file="org/apache/flink/streaming/api/operators/StreamOperator.html#finish--" name="StreamOperator#finish" >}}
> method. The method is expected to be a clear cutoff point for flushing any remaining buffered state.
> All checkpoints taken after the `finish` method has been called should be in most cases empty and
> shouldn't contain any buffered data, as there will be no way to emit this data. One notable
> exception is if your operator has some pointers to transactions in external systems, for example in
> order to implement the exactly-once semantic. In such a case, checkpoints taken after invoking `finish()`
> method should keep a pointer to the last transaction(s) that will be committed in the final checkpoint
> before the operator is closed. A good built-in example of this are exactly-once sinks and the 
> `TwoPhaseCommitSinkFunction`. What does it mean in more details?

Reviewer: ```suggestion
## Checkpointing with parts of the job graph finished *(BETA)*

Starting from Flink 1.14, it is possible to continue performing checkpoints even if parts of the job
graph have finished processing all the data. This might happen if there are bounded sources. This
feature must be enabled via a feature flag:

```java
Configuration config = new Configuration();
config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true);
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
```

Once the tasks/subtasks are finished, they do not contribute to the checkpoints any longer. This is an
an important consideration when implementing any custom operators or UDFs (User-Defined Functions). 

In order to support checkpointing with tasks that finished, we adjusted the [task lifecycle]({{< ref "docs/internals/task_lifecycle" >}}) and introduced the {{< javadoc file="org/apache/flink/streaming/api/operators/StreamOperator.html#finish--" name="StreamOperator#finish" >}} method. This method is expected to be a clear cutoff point for flushing any remaining buffered state. All checkpoints taken after the `finish` method has been called should be empty (in most cases) and should not contain any buffered data since there will be no way to emit this data. One notable exception is if your operator has some pointers to transactions in external systems (i.e. order to implement the exactly-once semantic). In such a case, checkpoints taken after invoking the `finish()` method should keep a pointer to the last transaction(s) that will be committed in the final checkpoint before the operator is closed. A good built-in example of this are exactly-once sinks and the `TwoPhaseCommitSinkFunction`. 
```

> ### Operator state
> 
> The aforementioned requirement implies certain contract especially on the operator state. First and
> foremost it forced us to implement a special handling for `UnionListState`. Often a `UnionListState`
> has been used to implement a global view over offsets in external system (e.g. we use it to store
> current offsets of Kafka partitions). If we had discarded a state for just a single subtask that had
> it's `finish` method called, we would've lost offsets for partitions it had been assigned. In order
> to work this problem around, we let checkpoints succeed, only if either none or all subtasks
> finished that use `UnionListState`.
> 
> We have not seen a `ListState` used in a similar way, but you must be aware that any state checkpointed
> after the `finish` method will be discarded and not available after a restore (if the corresponding
> checkpoint succeeds). We truly believe that any operator that is prepared to be rescaled should work
> well with tasks that partially finish (only a subset of it tasks finish). Restoring from a checkpoint
> which has some subtasks finished is equivalent to restoring such a task with the number of new
> subtasks equal to the number of finished tasks.

Reviewer: ```suggestion
### How does this impact the operator state?

There is a special handling for `UnionListState`, which has often been used to implement a global view over offsets in an external system (i.e. storing current offsets of Kafka partitions). If we had discarded a state for a single subtask that had its `finish` method called, we would have lost the offsets for partitions that it had been assigned. In order to work around this problem, we let checkpoints succeed only if none or all subtasks that use `UnionListState` are finished.

We have not seen `ListState` used in a similar way, but you should be aware that any state checkpointed after the `finish` method will be discarded and not be available after a restore. 

Any operator that is prepared to be rescaled should work well with tasks that partially finish. Restoring from a checkpoint where only a subset of tasks finished is equivalent to restoring such a task with the number of new subtasks equal to the number of finished tasks.
```

or UDFs. In order to support checkpointing with tasks that finish we adjusted the [task lifecycle]({{< ref "docs/internals/task_lifecycle" >}})
and introduced the {{< javadoc file="org/apache/flink/streaming/api/operators/StreamOperator.html#finish--" name="StreamOperator#finish" >}}
method. The method is expected to be a clear cutoff point for flushing any remaining buffered state.

> All checkpoints taken after the `finish` method has been called can contain only pointers to the last transactions
> that will be closed in the final checkpoint or in case of a failure should be closed when restoring.
> Apart from those pointers it should be empty. What does it mean in details?

Reviewer 2: ```suggestion
All checkpoints taken after the `finish` method has been called should be in most cases empty and
shouldn't contain any buffered data, as there will be no way to emit this data. One noteable 
exception is if your operator has some pointers to transactions in external systems, for example in 
order to implement exactly-once semantic. In such case checkpoints taken after invoking `finish()`
method should keep pointer to last transaction(s) that will be committed in the final checkpoint
before operator is closed. Good built-in example of this are exactly-once sinks and {{TwoPhaseCommitSinkFunction}}.
What does it mean in more details?
```

Reviewer 2: And what do you mean by:
> or in case of a failure should be closed when restoring.

? close -> committed? or aborted? I guess you mean committing when failure happens before `notifyCheckpointComplete()` reached the task? In such case I would skip this as too much details.
